### PR TITLE
Fix artifacts on rectangular lights

### PR DESCRIPTION
**Pull request by EvgeniiG**
### Purpose of this PR
Fix artifacts on rectangular lights.
Solve https://fogbugz.unity3d.com/f/cases/985076/

---
### Release Notes
Fix artifacts on rectangular lights

---
### Testing status
**Katana Tests**: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=HDRP%2Ffix-rect-lights&automation-tools_branch=master&unity_branch=trunk

---
### Overall Product Risks
**Technical Risk**: Low

**Halo Effect**: Low

### CODE REVIEW

**Comment by TomasKiniulis** [comment]
> For QA: Maybe we should add a test with various orientation for our rectangle area light. Current test for area light are all well oriented, face camera, above horizon. Let's do a test with random orientation, backface etc...

Created a favro ticket to add the test later

File: com.unity.render-pipelines.core/ShaderLibrary/AreaLighting.hlsl [hunk 1]
```diff
@@ -143,29 +143,18 @@ real3 PolygonFormFactor(real4x3 L)
...
 #ifdef APPROXIMATE_POLY_LIGHT_AS_SPHERE_LIGHT
-    UNITY_UNROLL
-    for (uint i = 0; i < 4; i++)
-    {
-        L[i] = normalize(L[i]);
-    }
+    if (cross(L[1] - L[0], L[2] - L[0]).z <= 0) return 0; // Light is back-facing
 
-    real3 F = real3(0, 0, 0);
+    if (max(max(max(L[0].z, L[1].z), L[2].z), L[3].z) <= 0) return 0; // Light is below the horizon
```
[the remark this answers is not in the document]
Comment (EvgeniiG) [reply com.unity.render-pipelines.core/ShaderLibrary/AreaLighting.hlsl]: The generated code will be the same, and this is somewhat cleaner, IMO.

File: com.unity.render-pipelines.core/ShaderLibrary/AreaLighting.hlsl [hunk 2]
```diff
@@ -143,29 +143,18 @@ real3 PolygonFormFactor(real4x3 L)
...
 
-        F += INV_TWO_PI * ComputeEdgeFactor(V1, V2);
-    }
+    real sff = saturate(dot(F, F));
 
-    // Clamp invalid values to avoid visual artifacts.
-    real f2         = saturate(dot(F, F));
-    real sinSqSigma = min(sqrt(f2), 0.999);
-    real cosOmega   = clamp(F.z * rsqrt(f2), -1, 1);
+    real formFactor        = sqrt(sff);
```
[the remark this answers is not in the document]
Comment (EvgeniiG) [reply com.unity.render-pipelines.core/ShaderLibrary/AreaLighting.hlsl]: Doesn't appear to be necessary anymore.

**Comment by EvgeniiG** [comment]
> Bug fix should target HDRP/Staging (but we can't change the target of this PR currently, so you will need to run Yamato).

Confusing messaging...

image

**Comment by sebastienlagarde** [comment]
Ah sorry. This message was general message for all package team. Saying that PR must target the main branch and and then backport. It was not for organisation inside hdrp team (thus the message in devs-renderpipe). Sorry for the confusion, I should have clarify

**Comment by EvgeniiG** [comment]
I'd like to add an additional emphasis that this must be thoroughly manually tested before merging to make sure there are no more leaks in any configurations.

**Comment by TomasKiniulis** [comment]
@EvgeniiG The edge light leak artifacts are fixed and I'm not getting any area light specular reflection viewing from the area lights back, but I'm getting new artifacts introduced with the pr - part of the light is not rendering viewing a bit from front side angle
9q3UJLPISr

**Comment by TomasKiniulis** [comment]
Attaching an angle comparison between staging and current PR.
Staging
image (18)
fix-rect-lights
image (17)

Adjusting plane smoothness value also affects broken angles

**Comment by EvgeniiG** [comment]
Weird, I will take a look.

**Comment by EvgeniiG** [comment]
OK, I can reproduce the problem.

**Comment by EvgeniiG** [comment]
Looks like there is a serious inherent issue with the sphere light approximation when the light is partially below the horizon...

**Comment by EvgeniiG** [comment]
I pushed another workaround, it's not perfect, but it appears to leak less. Please test!

**Comment by TomasKiniulis** [comment]
The previous issue seems resolved and the workaround does help with light leaks, although it looks like this also bends the light cast on the surface. Attaching Gif of the result. After the changes, when the light is moved to a certain point of height, you can see that the least light received is in the middle of the surface.
Current PR:
AyoZOVmnDK
hJ1U5pVwyX
Staging:
CpYmUEKxMK
iN5IO5t5f9

**Comment by EvgeniiG** [comment]
That's not really expected, but fixing this new issue may take a very long time. I would suggest we proceed with the current PR if it's an improvement over what we currently have.

**Comment by sebastienlagarde** [comment]
Hey, so I merge this PR. If there is a need for further investigation to solve the other issue mention in this thread, please proceed. Thanks (@EvgeniiG )

**Comment by EvgeniiG** [comment]
I pushed the fix.

**Pull request merged**